Entities/SceneLayer.h

```diff
@@ -1,544 +1,321 @@
+		/// </summary>
+		/// <param name="posX">The X coordinates of the position to wrap.</param>
+		/// <param name="posY">The Y coordinates of the position to wrap.</param>
+		/// <returns>Whether wrapping was performed or not.</returns>
+		bool WrapPosition(int &posX, int &posY) const;
+
+		/// <summary>
+		/// Only wraps a position coordinate if it is off bounds of the SceneLayer and wrapping in the corresponding axes are turned on.
+		/// </summary>
+		/// <param name="pos">The vector coordinates of the position to wrap.</param>
+		/// <returns>Whether wrapping was performed or not.</returns>
+		bool WrapPosition(Vector &pos) const { return ForceBoundsOrWrapPosition(pos, false); }
+
+		/// <summary>
+		/// Wraps or bounds a position coordinate if it is off bounds of the SceneLayer, depending on the wrap settings of this SceneLayer.
+		/// </summary>
+		/// <param name="posX">The X coordinates of the position to wrap.</param>
+		/// <param name="posY">The Y coordinates of the position to wrap.</param>
+		/// <returns>Whether wrapping was performed or not. Does not report on bounding.</returns>
+		bool ForceBounds(int &posX, int &posY) const;
+
+		/// <summary>
+		/// Wraps or bounds a position coordinate if it is off bounds of the SceneLayer, depending on the wrap settings of this SceneLayer.
+		/// </summary>
+		/// <param name="pos">The Vector coordinates of the position to wrap.</param>
+		/// <returns>Whether wrapping was performed or not. Does not report on bounding.</returns>
+		bool ForceBounds(Vector &pos) const { return ForceBoundsOrWrapPosition(pos, true); }
+#pragma endregion
+
+#pragma region Virtual Methods
+		/// <summary>
+		/// Updates the state of this SceneLayer.
+		/// </summary>
+		virtual void Update() {}
+
+		/// <summary>
+		/// Draws this SceneLayer's current scrolled position to a bitmap.
+		/// </summary>
+		/// <param name="targetBitmap">The bitmap to draw to.</param>
+		/// <param name="targetBox">The box on the target bitmap to limit drawing to, with the corner of box being where the scroll position lines up.</param>
+		/// <param name="offsetNeedsScrollRatioAdjustment">Whether the offset of this SceneLayer or the passed in offset override need to be adjusted to scroll ratio.</param>
+		virtual void Draw(BITMAP *targetBitmap, Box &targetBox, bool offsetNeedsScrollRatioAdjustment = false);
+#pragma endregion
+
+	protected:
+
+		static Entity::ClassInfo m_sClass; //!< ClassInfo for this class.
+
+		ContentFile m_BitmapFile; //!< ContentFile containing the path to this SceneLayer's sprite file.
+		BITMAP *m_MainBitmap; //!< The BITMAP of this SceneLayer.
+		bool m_MainBitmapOwned; //!< Whether the main bitmap is owned by this.
+		bool m_DrawMasked; //!< Whether pixels marked as transparent (index 0, magenta) are skipped when drawing or not (masked drawing).
+
+		bool m_WrapX; //!< Whether wrapping is enabled on the X axis.
```

Comment:
SceneLayer is the base class for both of them. SLTerrain is just a collection of SceneLayers.

Reply:
Did the terrain stop scrolling and wrapping or something?

Reply:
Yes? I dunno, my comment was 4 months ago, I guess I'd thought it didn't scroll at least?

Reply:
lel
SceneLayer handles all the scrolling and wrapping for the terrain (which inherits from SL and stores another 2 SLs that just copy the resulting offset after an update).
SLBG just handles animation and autoscroll and other misc BG specific junk, the scrolling and wrapping is still handled in SL.

Reply:
Okay, if terrain needs scrolling as well then obviously it needs to stay where it is. Maybe I was thinking that you had stuff for autoscrolling in SL instead of SLBG?
Anyway, if this is literally how things need to be in order to work, that's fine, feel free to resolve.

Reply:
Autoscrolling is exclusively a SLBG thing, terrain does not support autoscroll and neither does SL itself.
Don't confuse scrolling with autoscrolling.